arbnode/espresso_force_inclusion_checker.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		return 0, err
+	}
+	log.Debug("Max time variation delay blocks", "maxTimeVariationDelayBlocks", maxTimeVariationDelayBlocks, "maxTimeVariationDelaySeconds", maxTimeVariationDelaySeconds)
+
+	parentLatestHeader, err := f.l1Reader.LastHeader(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	l1BlockNumber := parentLatestHeader.Number.Uint64()
+	l1TimeStamp := parentLatestHeader.Time
+
+	log.Debug("L1 block number", "l1BlockNumber", l1BlockNumber, "l1TimeStamp", l1TimeStamp)
+
+	if f.l1Reader.IsParentChainArbitrum() {
+		headerInfo := types.DeserializeHeaderExtraInformation(parentLatestHeader)
+		l1BlockNumber = headerInfo.L1BlockNumber
+	}
+
+	lastBadBlockNumber := arbmath.SaturatingUSub(f.config.BlockThresholdTolerance+l1BlockNumber, arbmath.BigToUintSaturating(maxTimeVariationDelayBlocks))
+	lastBadBlockTime := arbmath.SaturatingUSub(f.config.SecondThresholdTolerance+l1TimeStamp, arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
+
+	log.Debug("Last bad block number", "lastBadBlockNumber", lastBadBlockNumber, "lastBadBlockTime", lastBadBlockTime)
+
+	if f.l1Reader.IsParentChainArbitrum() {
+		n, err := node_interfacegen.NewNodeInterface(types.NodeInterfaceAddress, f.l1Reader.Client())
+		if err != nil {
+			return 0, err
+		}
+		rng, err := n.L2BlockRangeForL1(&bind.CallOpts{Context: ctx}, lastBadBlockNumber)
+		if err == nil {
+			lastBadBlockNumber = rng.LastBlock
+			log.Debug("Last bad block number from L2 block range for L1", "lastBadBlockNumber", lastBadBlockNumber)
+		} else {
+			// If the L2 block range for L1 call fails, we will use binary search
+			// The start block number should be the genesis block
+			// Reference: https://github.com/OffchainLabs/arbitrum-sdk/blob/792a7ee3ccf09842653bc49b771671706894cbb4/src/lib/inbox/inbox.ts#L104-L113
+			genesis, err := n.NitroGenesisBlock(&bind.CallOpts{Context: ctx})
+			if err != nil {
+				return 0, err
+			}
+			target := lastBadBlockNumber
+			start := genesis.Uint64()
+			end := parentLatestHeader.Number.Uint64()
+			lastBadBlockNumber, err = binarySearchForBlockNumber(ctx, start, end, func(ctx context.Context, blockNumber uint64) (int, error) {
+				block, err := f.l1Reader.Client().BlockByNumber(ctx, arbmath.UintToBig(blockNumber))
+				if err != nil {
+					return 0, err
+				}
+				l1Block := types.DeserializeHeaderExtraInformation(block.Header()).L1BlockNumber
+				if l1Block < target {
+					return binarySearch_LessThanTarget, nil
+				} else if l1Block > target {
+					return binarySearch_GreaterThanTarget, nil
+				} else {
+					return binarySearch_EqualToTarget, nil
+				}
+			})
+			if err != nil {
+				log.Error("error in binary search", "err", err)
+				return 0, err
+			}
+			log.Debug("Last bad block number from binary search", "lastBadBlockNumber", lastBadBlockNumber)
+		}
+	}
+
+	lastBadBlock, err := f.findFirstParentChainBlockBelow(ctx, lastBadBlockNumber, lastBadBlockTime)
```

Comment:
Can you please explain why we are simply calling this function if the parent chain is not arbitrum but Ethereum?

Reply:
The calculation of the force inclusion window is always based on `ethereum`, no matter the parent chain is `etherrum` or `arbitrum`. So if the parent chain is an L2, we have to calculate the corresponding L1 blocks.

Here is the PR https://github.com/EspressoSystems/nitro-espresso-integration/pull/616

And you can check the related asana tasks.